docs/guides/admin/docs/configuration/security.jwt.md

```diff
@@ -1,18 +1,107 @@
+
+This schema is designed to be flexible and support all current and conceivable future use cases, while keeping the JWT size for each use case small.
+
+### Quick overview
+
+```json5
+{
+    // Standard claims
+    "exp": 1499863217, // required
+    "nbf": 1548632170, // optional
+
+    // User info
+    "username": "jose",
+    "name": "José Carreño Quiñones",
+    "email": "jose@example.com",
+
+    // Authorization, i.e. privileges the request/user should have
+    "roles": ["STUDIO", "API_EVENTS_VIEW"], // without `ROLE_` prefix
+    "oc": [
+        "read+customaction:e:d622b861-4264-4947-8db1-c754c5956433",
+        "write:s:4ed02421-144c-42a1-b98a-22e84f3ac691"
+    ]
+}
+```
+
+### Claims
+
+A claim being "required" means that the JWT provider (e.g. Tobira, LMS) has to include it. Opencast will reject all JWTs which lack required claims.
+
+- `exp` (expiration time): As [defined in the JWT RFC](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4), a number timestamp in seconds since the unix epoch. JWTs with `exp < now()` are rejected by Opencast. This claim is *required*.
+
+- `nbf` (not before): As [defined in the JWT RFC](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.5), a number timestamp in seconds since the unix epoch. JWTs with `nbf > now()` are rejected by Opencast. This claim is optional.
+
+- *User Info*: information about the user that's possibly used by parts of Opencast. Some JWT use cases don't require user information at all (e.g. static file auth), while others do (e.g. Studio). Therefore, these claims are optional. (TODO: specify when this is stored as user reference)
+    - `username`: string. (TODO: explain how this is used to lookup existing users, if at all)
+    - `name`: display name, string.
+    - `email`: string.
+
+- *Authorization*: at least one of these claims has to be set, as otherwise no privileges are granted (which is the same as not sending a JWT at all).
+    - `roles`: array of strings, specifying roles to grant to the request/user directly. The `ROLE_` prefix must be omitted as Opencast adds it itself.
+    - `oc`: array of strings, grants access to single items in Opencast. The custom syntax is terse in order to keep JWTs as small as possible. Each string consists of three parts, which are separated by a colon (`:`):
+        - *Action(s)*: the first part specifies what actions are granted, corresponding to actions in the OC ACLs, e.g. `read` and `write`. Multiple actions can be specified by joining them with `+`. To keep the syntax unambigious, actions cannot contain `:` or `+` (ideally they are ASCII alphabetic only!). To further preserve space, some actions imply others:
```

Comment:
Know actions include `read`, `write`, `annotate` and `annotate-admin`. I think TU Vienna uses some custom actions and I'm defining one as well. I don't think we restrict the action name currently, but at the same time, it's probably unlikely that `+` and `:` is used.

Reply:
Yeah, that was my thinking: I don't know any actions that use these two characters. We can always define some escape-mechanism later, if people really need to.

Reply:
Wouldn't it be easier to deal with a claim structured in the following way later on?

```
"entitlements" : {"eventid-1":["read","write"],"eventid-2":["read","write", "annotate"]}
"roles" : ["studio","editor"]
```

[entitlements claim -> see RFC 7643](https://www.rfc-editor.org/rfc/rfc7643.html)

Reply:
<sup>Sorry for late reply, I was sick</sup>

`entitlements` is very interesting and I didn't know about that claim. It does indeed sound like pretty much what I described with my `oc` claim. Of course, `oc` is shorter than `entitlements`, I'm not sure whether it's worth ignoring standards for some JWT size improvements (aside: I hate that this is even something we have to think about, a factor to weight against, exchange formats shouldn't be designed to be human readable, ramble ramble ramble...).

The structure of `entitlements` you suggest is also interesting. Again, slightly larger than my suggestion but probably insignificantly so. 

> Do we need to differentiate between series, playlists and episodes? Or is the uuid sufficiant?

I would argue the "type" is important. Sure, UUIDs rarely clash, but remember that IDs can be given manually and you don't want accidentally also give access to some totally unrelated item. Your "JSON object" structure could just be extended by using `e:uuid1` as key names.

Finally: `groups`. Not sure if we want to add this as well. Yes Opencast has groups, but these don't have a particularly nice identifier. You can always give the `ROLE_GROUP_` role instead to have the same effect. And we can always add this feature later. So I'd suggest ignoring it for now, to be able to merge this faster.

---

So, for me, `entitlements` and `oc` are both fine. Same for the structure: my array or your object (with the type of item added to the key) are both fine. Your suggestions are probably slightly nicer in both cases. However, they make the JWT larger and I don't have any context for RFC 7643, so not sure how useful that is.

I will bring this up in the technical meeting tomorrow.

Reply:
If we want to follow RFC 7643, the suggested structure is as follows:

```json
"entitlements": [
  {
    "value": "entitlement-1",
    "display": "Entitlement 1 gives some rights",
    "type": "entitlement-1-function",
    "primary": false
  }
]
```

All ``
value
display

Reply:
> If we want to follow RFC 7643, the suggested structure is as follows:

This seems like an implementation suggestion. As all properties are listed as _required: false_ within the scheme, you are basically allowed to extend the scheme of entitlements. The definition for entitlements itself says:
 - "No vocabulary or syntax is specified; [...] service providers and clients are expected to encode sufficient information in the value [...] This value has no canonical types, although a type may be useful as a means to scope entitlements." ([Source](https://datatracker.ietf.org/doc/html/rfc7643))

I think a type could be a custom "OC" Resource Schema, comparable to e.g. the [group resource schema.](https://www.rfc-editor.org/rfc/rfc7643#section-4.2), defining kind of an oc class as the rfc states that it provides a  schema and extension model for representing users and groups and **_other resource types_** in JSON format, also gives hints to "defining new resource types".

So with the scheme proposed here originally:
```
"oc": [
        "read+customaction:e:d622b861-4264-4947-8db1-c754c5956433",
        "write:s:4ed02421-144c-42a1-b98a-22e84f3ac691"
    ]
```
following the rfc implementation suggestion:
```
  "entitlements": [
        {"value": "read+customaction:e:d622b861-4264-4947-8db1-c754c5956433"},
        {"value": "write:s:4ed02421-144c-42a1-b98a-22e84f3ac691"}
    ]
```
That might not be too far off, although I'd recommend that the claim should be as easily processable as possible, e.g. I'd think of json properties instead of needing to split strings by colons or something, so with an "OC" resource type schema, e.g.

```
"entitlements" : [
        {"uuid-1":["read","write"]},
        {"uuid-2":["read","write", "annotate"]}
]
```
or some more complex, e.g.
```
  "entitlements": [
        {"value": "d622b861-4264-4947-8db1-c754c5956433","acl": ["read", "customaction"],"type":["e"]},
       {"value": "4ed02421-144c-42a1-b98a-22e84f3ac691","acl": ["write"],"type":["s"]},
 ]
```
Json objects instead of strings should also allow for easier future modifications, since properties could be added more easily.

> I would argue the "type" is important. Sure, UUIDs rarely clash, but remember that IDs can be given manually and you don't want accidentally also give access to some totally unrelated item. Your "JSON object" structure could just be extended by using e:uuid1 as key names.

Is this a real use case? Aren't UUIDs of an OC instance always unique? Or is the uniqueness only assured by types? There is probably also a uniqueness check for manual IDs?

Reply:
> Is this a real use case? Aren't UUIDs of an OC instance always unique? Or is the uniqueness only assured by types? There is probably also a uniqueness check for manual IDs?

No they are not. I justed created a series and an event in Opencast, both with the id `foo`. So yes, uniqueness is only ensured if we add the type to the ID.

---

Regarding the data structure discussion: I would absolutely not go for a verbose structure like quote by Matthias. Even if most fields are optional, just having the same field name in every JWT is just such a waste. I like the argument of using standard JSON to avoid custom parsing of strings. Again, I'm in favor of either my initial suggestion or, slightly preferred, the first one by @oellers (the json object with uuids+type as keys). 

> Wouldn't it then make sense to split up those ACLs instead of combining with e.g. "plugs"?

The reason I suggested this is that granting multiple "actions" to the same item is common and we want to avoid repeating the ID multiple times.

Reply:
We discussed this in the technical meeting just now and decided:

- To use `oc` instead of `entitlements`: using the latter is only useful if we plan on interfacing with any systems using the linked RFC. This is very unlikely and we cannot think of any reasonable situation where we might. Further, the RFC suggests (yes, only suggests, but still) a schema for that claim, which we wouldn't use anyway. So only half following the RFC seems not useful.
- To use the `{ "e:uuid": ["read"] }` syntax for the `oc` claim instead of the array one I proposed. Using proper JSON syntax instead of custom string syntax avoids writing custom parsers (even if they are super simple), side steps the problem of having actions (e.g. `read`) containing colons or plus signs (which would require escaping), and also makes this a bit more flexible as we could add other fields to the object in the future.

I will update this PR soon.